File: src/main/java/com/cobblegyms/util/TimeUtil.java
package com.cobblegyms.util;

import java.time.*;
import java.time.temporal.IsoFields;
import java.time.temporal.WeekFields;

public class TimeUtil {

    public static String formatDuration(long millis) {
        if (millis <= 0) return "0 seconds";

        long seconds = millis / 1000;
        long minutes = seconds / 60;
        long hours = minutes / 60;
        long days = hours / 24;

        seconds %= 60;
        minutes %= 60;
        hours %= 24;

        StringBuilder sb = new StringBuilder();
        if (days > 0) sb.append(days).append(days == 1 ? " day" : " days");
        if (hours > 0) {
            if (!sb.isEmpty()) sb.append(", ");
            sb.append(hours).append(hours == 1 ? " hour" : " hours");
        }
        if (minutes > 0) {
            if (!sb.isEmpty()) sb.append(", ");
            sb.append(minutes).append(minutes == 1 ? " minute" : " minutes");
        }
        if (seconds > 0 && days == 0 && hours == 0) {
            if (!sb.isEmpty()) sb.append(", ");
            sb.append(seconds).append(seconds == 1 ? " second" : " seconds");
        }

        return sb.isEmpty() ? "0 seconds" : sb.toString();
    }

    public static long getWeekStart() {
        LocalDate today = LocalDate.now(ZoneOffset.UTC);
        LocalDate monday = today.with(WeekFields.ISO.dayOfWeek(), 1);
        return monday.atStartOfDay(ZoneOffset.UTC).toInstant().toEpochMilli();
    }

    public static boolean isExpired(long timestamp) {
        return System.currentTimeMillis() > timestamp;
    }

    public static String formatTimestamp(long timestamp) {
        Instant instant = Instant.ofEpochMilli(timestamp);
        ZonedDateTime zdt = instant.atZone(ZoneOffset.UTC);
        return String.format("%04d-%02d-%02d %02d:%02d UTC",
                zdt.getYear(), zdt.getMonthValue(), zdt.getDayOfMonth(),
                zdt.getHour(), zdt.getMinute());
    }

    public static long hoursToMillis(int hours) {
        return (long) hours * 3600 * 1000;
    }
}

File: src/main/kotlin/com/cobblegyms/util/TimeUtil.kt
package com.cobblegyms.util

import java.time.Instant
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter

object TimeUtil {
    
    private val DATE_FORMAT = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm")
        .withZone(ZoneOffset.UTC)
    
    fun nowSeconds(): Long = System.currentTimeMillis() / 1000
    
    fun formatTimestamp(epochSeconds: Long): String {
        return DATE_FORMAT.format(Instant.ofEpochSecond(epochSeconds))
    }
    
    fun formatDuration(seconds: Long): String {
        val days = seconds / 86400
        val hours = (seconds % 86400) / 3600
        val minutes = (seconds % 3600) / 60
        val secs = seconds % 60
        
        return buildString {
            if (days > 0) append("${days}d ")
            if (hours > 0) append("${hours}h ")
            if (minutes > 0) append("${minutes}m ")
            if (days == 0L && hours == 0L && minutes == 0L) append("${secs}s")
        }.trim()
    }
    
    fun getWeekStartSeconds(): Long {
        val cal = java.util.Calendar.getInstance()
        cal.set(java.util.Calendar.DAY_OF_WEEK, java.util.Calendar.MONDAY)
        cal.set(java.util.Calendar.HOUR_OF_DAY, 0)
        cal.set(java.util.Calendar.MINUTE, 0)
        cal.set(java.util.Calendar.SECOND, 0)
        cal.set(java.util.Calendar.MILLISECOND, 0)
        return cal.timeInMillis / 1000
    }
    
    fun isOnCooldown(lastTime: Long, cooldownHours: Int): Boolean {
        val elapsed = nowSeconds() - lastTime
        return elapsed < cooldownHours * 3600L
    }
    
    fun getCooldownRemaining(lastTime: Long, cooldownHours: Int): Long {
        val elapsed = nowSeconds() - lastTime
        return maxOf(0L, cooldownHours * 3600L - elapsed)
    }
}
